proposals/0002-monitoring.md

```diff
@@ -0,0 +1,178 @@
+# Monitoring via HTTP endpoints
+
+* Proposal: [MP-0002](0002-monitoring.md)
+* Authors: [Heiko W. Rupp](https://github.com/pilhuhn)
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!forum/microprofile), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+To ensure reliable operation of the software it is necessary to monitor essential
+system parameters. This enhancement proposes the addition of well-known monitoring
+endpoints for each process adhering to the microprofile standard.
+
+This proposal does not talk about health checks. There will be a separate proposal.
+
+Mailinglist thread: [Monitoring and Healthcheck requirements](https://groups.google.com/forum/#!topic/microprofile/jIZAKiu76ys)
+
+## Motivation
+
+Reliable service of a platform needs monitoring. There is already JMX,
+but remote-JMX is not easy to deal with and especially does not fit well
+in a polyglott environment where other services are not running on the 
+JVM.
+To enable monitoring in an easy fashion it is necessary that all microprofile
+implementations follow a certain standard with respect to (base) API path,
+data types involved and return codes used.
+
+## Proposed solution
+
+Data is exposed via REST over http under the `/metrics` base path
+
+* Metrics will respond to GET requests and are exposed in a tree like fashion with sub-trees for various cases described below.
+* A 'shadow tree' that responds to OPTIONS will provide the metadata                 
+* All data is at least exposed as application/Json data type.
+* The api is discoverable. GETting a base object will list references to children
+
+Data access must honour the http response codes, especially
+
+* 200 for successful retrieval of an object
+* 204 when retrieving a subtree that would exist, but has no content. E.g. when the application-specific subtree has no application specific metrics defined.
+* 404 if an directly-addressed item does not exist. This may be a non-existing sub-tree or non-existing object
+* 503 to indicate that a health check resulted in "bad health". The body SHOULD contain details { "details": <text> }
```

Comment:
503 means "Service Unavailable", wondering if this is the right http status as sometimes we could have a very high CPU usage which means bad health but not unavailable.